eluthia/zipapp_script.py

```diff
@@ -0,0 +1,187 @@
+    Input:
+        string zip_path - path to the archive to be extracted (this should always be os.path.dirname(__file__), but it's a parameter for modularity)
+        string unpack_path - path to where the packages should be placed
+    Output:
+        list package_list - list of packages inside the archive.
+    """
+    with zipfile.ZipFile(zip_path) as z:
+        package_list = [item for item in z.namelist() if item.lower().endswith('.deb')]
+
+        for package in package_list:
+            z.extract(package, unpack_path)
+
+    return package_list
+
+
+def repack(package_dict, repack_path):
+    """
+    Repackage installed debian packages into .deb files
+    Input:
+        dict package_dict - a dictionary of packages to repack - Format: {package_name: version}
+        string repack_path - path to where repackaged .deb files should be placed
+    """
+    if get_version('dpkg_repack') is None:
+        subprocess.run(['apt-get', 'install', 'dpkg-repack'])
+
+    for package_name, version in package_dict.items():
+        subprocess.run(['dpkg-repack', package_name], cwd=repack_path)
+        os.rename(f'{repack_path}/{package_name}_{version}_all.deb', f'{package_name}.deb')
+
+    return
+
+
+def verify_installation(packages, preinst_file_count, history_folder, timeout, check_frequency):
+    """
+    Verifies that packages are installed and working correctly.
+    Input:
+        list packages - new packages to be installed
+        int preinst_file_count - Number of files that were in badtrack's history folder before installation.
+        string history_folder - path to badtrack's history folder
+        int timeout - how long to give badtrack to make it's first history file before rolling back.
+        int check_frequency - how often to check the history folder for new files.
+    """
+    def verify_badtrack(preinst_file_count, history_folder, timeout, check_frequency):
+        stop_time = time.time() + timeout
+        while time.time() < stop_time: # Verify that badtrack is making files
+            time.sleep(check_frequency)
+            if len(os.listdir(history_folder)) > preinst_file_count:
+                return True
+        return False
+
+    missing_packages = [package for package in packages if get_version(package) == None] # packages in new_packages that aren't installed
+
+    if missing_packages:
+        raise InstallationError(f'Packages were not installed: {missing_packages}')
+
+    if not verify_badtrack(preinst_file_count, history_folder, timeout, check_frequency):
+        raise InstallationError('No files were generated by badtrack within 1 minute')
+
+    return
+
+
+def rollback(new_packages, old_packages):
+    """
+    Rolls back the installation of packages.
+    Input:
+        list new_packages - installed packages to be replaced
+        list old_packages - paths to old debian files to replace the installed versions
+    """
+    subprocess.run(['dpkg', '-r'] + list(new_packages))
```

Comment:
Spell `-r` out to `--remove`.

Reply:
Remove `list` typecasting here. I would typecast `new_packages` to a list, before calling `rollback` function instead:
```py
rollback(list(NEW_PACKAGES.keys()), ...)
```

Reply:
Need `check=True` argument for subprocess.run.

Reply:
Perhaps we should create our own `run_command` function that always sets `check=True`.